Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp

```diff
@@ -0,0 +1,249 @@
+            terrainSrg->SetImage(m_detailMiscClipmapsIndex, m_clipmaps.at(AZ::Name(DetailMiscClipmapString)));
+
+            m_clipmapsNeedUpdate = false;
+        }
+    }
+
+    void TerrainClipmapManager::InitializeClipmapData()
+    {
+        float clipmapScaleInv = 1.0f;
+        for (int32_t clipmapIndex = ClipmapStackSize - 1; clipmapIndex >= 0; --clipmapIndex)
+        {
+            AZ::Vector4(0.5f).StoreToFloat4(m_clipmapData.m_clipmapCenters[clipmapIndex].data());
+            AZ::Vector4(clipmapScaleInv).StoreToFloat4(m_clipmapData.m_clipmapScaleInv[clipmapIndex].data());
+            clipmapScaleInv /= 2.0f;
+        }
+
+        AZ::Vector4::CreateZero().StoreToFloat4(m_clipmapData.m_viewPosition.data());
+
+        m_clipmapData.m_clipmapSize[0] = ClipmapSizeWidth;
+        m_clipmapData.m_clipmapSize[1] = ClipmapSizeHeight;
+
+        AZ::Aabb worldBounds = AZ::Aabb::CreateNull();
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            worldBounds, &AzFramework::Terrain::TerrainDataRequests::GetTerrainAabb);
+
+        m_clipmapData.m_worldBounds[0] = worldBounds.GetMin().GetX();
+        m_clipmapData.m_worldBounds[1] = worldBounds.GetMin().GetY();
+        m_clipmapData.m_worldBounds[2] = worldBounds.GetMax().GetX();
+        m_clipmapData.m_worldBounds[3] = worldBounds.GetMax().GetY();
+
+        // Use world size for now.
+        const AZ::Vector3 worldSize = worldBounds.GetMax() - worldBounds.GetMin();
+        AZ::Vector2(worldSize.GetX(), worldSize.GetY()).StoreToFloat2(m_clipmapData.m_maxRenderSize.data());
+    }
+
+    void TerrainClipmapManager::InitializeClipmapImages()
+    {
+        AZ::Data::Instance<AZ::RPI::AttachmentImagePool> pool = AZ::RPI::ImageSystemInterface::Get()->GetSystemAttachmentPool();
+
+        AZ::RHI::ImageDescriptor imageDesc;
+        imageDesc.m_format = AZ::RHI::Format::R8G8B8A8_UNORM;
+        imageDesc.m_bindFlags = AZ::RHI::ImageBindFlags::ShaderReadWrite;
+        imageDesc.m_size = AZ::RHI::Size(ClipmapSizeWidth, ClipmapSizeHeight, 1);
+        imageDesc.m_arraySize = ClipmapStackSize;
+
+        AZ::Name macroColorClipmapName = AZ::Name(MacroColorClipmapString);
+        AZ::Data::Instance<AZ::RPI::AttachmentImage> macroColorClipmaps =
+            AZ::RPI::AttachmentImage::Create(*pool.get(), imageDesc, macroColorClipmapName, nullptr, nullptr);
+        AZ::Name detailColorClipmapName = AZ::Name(DetailColorClipmapString);
+        AZ::Data::Instance<AZ::RPI::AttachmentImage> detailColorClipmaps =
+            AZ::RPI::AttachmentImage::Create(*pool.get(), imageDesc, detailColorClipmapName, nullptr, nullptr);
+        AZ::Name detailMiscClipmapName = AZ::Name(DetailMiscClipmapString);
+        AZ::Data::Instance<AZ::RPI::AttachmentImage> detailMiscClipmaps =
+            AZ::RPI::AttachmentImage::Create(*pool.get(), imageDesc, detailMiscClipmapName, nullptr, nullptr);
+
+        imageDesc.m_format = AZ::RHI::Format::R8G8B8A8_SNORM;
```

Comment:
Should we consider a different format to store normals? I'm curious if you noticed any artifacts here since I imagine this will be a pretty lossy conversion from an artist-authored detail normal map. Maybe R16B16 (with implicit 3rd component) would be better for example.

Reply:
Yeah. I'll change to R16G16. I was thinking if we can just calculate everything ahead of the forward pass, so that we don't maintain 2 different sampling functions from rendering using clipmaps or using materials directly.